[FSDP][Collectives] skipping allgather when world size is 1

anshul-si opened the pull request:
**Summary:** In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated fsdp_params group to skip the foreach_all_gather and foreach_all_gather_copy_out APIs when world_size ‎ = 1. I have created a test that uses CommDebugMode to verify that the all gather comm has been removed. I also edited an affected test which used 1-way FSDP by verifying and changing its assert statements for CommDebugMode. Below, I have included the link to the profile trace verifying these two APIs were skipped and two test commands. 

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_f846ac3b-9467-4060-8e36-8cc3bc4449c3_devgpu263.prn2.facebook.com_652183.1753822140871934814.pt.trace.json

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160457
* #160443
* #160441
* #160390
* #160344
* #160147
* #160136
* __->__ #160135



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

anshul-si committed:
[FSDP][Collectives] skipping allgather when world size is 1

[ghstack-poisoned]

anshul-si committed:
Update on "[FSDP][Collectives] skipping allgather when world size is 1"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

weifengpy commented on torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:
```diff
@@ -336,25 +353,60 @@ def wait_for_unshard(self):
+                fsdp_param.init_all_gather_outputs(
+                    [all_gather_input.numel()],
+                    [all_gather_input.dtype],
+                    1,
```
nit: use world_size for better readability

weifengpy commented:
attach a trace showing all-gather are indeed skipped and there is only copy_ overhead?

anshul-si committed:
Update on "[FSDP][Collectives] skipping allgather when world size is 1"

**Summary:** In its current state, FSDP collectives uses cuda synchronizations and communication ops regardless of what the world size is. However, now that replicate will use FSDP, there will be instances where group size = 1 and these synchronizations and ops will be used needlessly. I have updated fsdp_params group to skip the foreach_all_gather and foreach_all_gather_copy_out APIs when world_size ‎ = 1. I have created a test that uses CommDebugMode to verify that the all gather comm has been removed. I also edited an affected test which used 1-way FSDP by verifying and changing its assert statements for CommDebugMode. Below, I have included the link to the profile trace verifying these two APIs were skipped and two test commands. 

https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/anshulsi_f846ac3b-9467-4060-8e36-8cc3bc4449c3_devgpu263.prn2.facebook.com_652183.1753822140871934814.pt.trace.json




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/test_composability/test_2d_composability.py:
```diff
@@ -277,19 +277,19 @@ def test_tp_with_fsdp_offloading(self):
 
             fwd_comm_counts = fwd_comm_mode.get_comm_counts()
-            self.assertEqual(len(fwd_comm_counts), 2)
+            self.assertEqual(len(fwd_comm_counts), 1)
```
I missed this: I missed this unit test. what's that fwd_comm_counts, bwd_comm_counts changed?

anshul-si replied:
Since I skip over foreach_gather and foreach_gather_copyout when world_size is 1, the all_gather count drops to 0, which is also why the number of different entries in fwd_comm_counts dict becomes 1. Same thing happens in the backward pass

weifengpy replied:
got you.  thanks for clarifying
